File: 143-reorder-list/reorder-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
var reorderList = function(head) {
    if (!head) return

    let mid = getMid(head)
    secondHead = mid.next
    mid.next = null

    let reversedMid = reverseLinkedList(secondHead)

    mergeLists(head, reversedMid)
};

function getMid(head) {
    if (head === null) return head

    let fast = head
    let slow = head

    while (fast.next && fast.next.next) {
        fast = fast.next.next
        slow = slow.next
    }

    return slow
}

function reverseLinkedList(head) {
    if (!head) return head

    let prev = null
    let next = null
    let current = head

    while (current) {
        next = current.next
        current.next = prev
        prev = current
        current = next
    }

    return prev
}

function mergeLists(first, second) {
    let p1 = first
    let p2 = second

    while (p2) {
        let temp1 = p1.next
        let temp2 = p2.next

        p1.next = p2
        p2.next = temp1

        p1 = temp1
        p2 = temp2
    }
}

// 1 -> 2
// 4 -> 3

/*
temp1 = 2
temp2 = 3

p1 = 1 -> 4 -> 2
p2 = 4 -> 2 -> 3

*/

File: 143-reorder-list/reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if not head or not head.next or not head.next.next:
         return  # Lists with 0, 1, or 2 nodes don't need reordering
        
        # Step 1: Find the middle of the linked list
        slow = fast = head
        while fast and fast.next:
            slow = slow.next       # Moves one step at a time
            fast = fast.next.next  # Moves two steps at a time
        
        # Now 'slow' points to the middle node
        
        # Step 2: Reverse the second half of the linked list
        second_half = slow.next  # Start of second half
        slow.next = None         # Break the list into two halves
        
        # Reverse the second half
        prev = None
        current = second_half
        while current:
            next_temp = current.next  # Store next node
            current.next = prev       # Reverse the pointer
            prev = current            # Move prev forward
            current = next_temp       # Move current forward
        
        reversed_second_half = prev   # New head of the reversed second half
        
        # Step 3: Merge the two halves in alternating fashion
        first = head
        second = reversed_second_half
        
        while second:  # Second half could be shorter (or equal) if list has odd length
            # Save next pointers
            first_next = first.next
            second_next = second.next
            
            # Interweave the nodes
            first.next = second
            second.next = first_next
            
            # Move to the next pair of nodes
            first = first_next
            second = second_next